sbncode/BeamSpillInfoRetriever/BNBRetriever/BNBRetriever_module.cc

```diff
@@ -0,0 +1,434 @@
+
+    // Iterating through each of the beamline times
+    for (size_t i = 0; i < times_temps.size(); i++) {
+
+      // Only continue if these times are matched to our DAQ time
+      // plus or minus some time padding, currently using 3.3 ms 
+      // which is half the Booster Rep Rate
+      
+      if(e.event() != 1){//We already addressed the "first event" above
+	if(times_temps[i] > (t_current_event+fTimePad)){continue;}
+	if(times_temps[i] <= (t_previous_event-fTimePad)){continue;}
+      }
+
+      //Loop through the multiwire devices:
+      
+      for(int dev = 0; dev < int(MWR_times.size()); dev++){
+
+	//Loop through the multiwire times:
+	double Tdiff = 1000000000.;
+	matched_MWR[dev] = 0;
+	bool best_match = false;
+	for(int mwrt = 0;  mwrt < int(MWR_times[dev].size()); mwrt++){
+	  
+	  //found a candidate match! 
+	  if(fabs((MWR_times[dev][mwrt] - times_temps[i])) < Tdiff){
+	    best_match = true;
+	    
+	    //Check for a better match...
+	    for (size_t j = 0; j < times_temps.size(); j++) {
+	      if( j == i) continue;
+	      if(times_temps[j] > (t_current_event+fTimePad)){continue;}
+	      if(times_temps[j] <= (t_previous_event-fTimePad)){continue;}
+	      
+	      //is there a better match later in the spill sequence
+	      if(fabs((MWR_times[dev][mwrt] - times_temps[j])) < 
+		 fabs((MWR_times[dev][mwrt] - times_temps[i]))){
+		//we can have patience...
+		best_match = false;
+		break;
+	      }	     
+	    }//end better match check
+
+	    //Verified best match!
+	    if(best_match == true){
+	      matched_MWR[dev] = mwrt;
+	      Tdiff = fabs((MWR_times[dev][mwrt] - times_temps[i]));
+	    }
+	  }//Find matches between 
+	}//end loop over MWR times 
+	
+      }//end loop over MWR devices
+
+
+      //Great we found a matched spill! Let's count it
+      spill_count++;
+      
+      // initializing all of our device carriers
+      // device definitions can be found in BNBSpillInfo.h
+
+      double TOR860 = 0; // units e12 protons
+      double TOR875 = 0; // units e12 protons
+      double LM875A = 0; // units R/s
+      double LM875B = 0; // units R/s
+      double LM875C = 0; // units R/s
+      double HP875 = 0; // units mm
+      double VP875 = 0; // units mm
+      double HPTG1 = 0; // units mm
+      double VPTG1 = 0; // units mm
+      double HPTG2 = 0; // units mm
+      double VPTG2 = 0; // units mm
+      double BTJT2 = 0; // units Deg C
+      double THCURR = 0; // units kiloAmps
+      
+      double TOR860_time = 0; // units s
+      
+      // Here we request all the devices
+      // since sometimes devices fail to report we'll
+      // allow each to throw an exception but still move forward
+      // interpreting these failures will be part of the beam quality analyses 
+      try{bfp->GetNamedData(times_temps[i], "E:TOR860@",&TOR860,&TOR860_time);}catch (WebAPIException &we) {std::cout << "At time : " << times_temps[i] << " " << "got exception: " << we.what() << "\n";}
```

Comment:
Please put this into a function...
```suggestion
      auto fillFromDev = [this](double time, std::string const& varList, auto&... dest)
        {
          try { bfp->GetNamedData(time, varList, &dest...); }
          catch (WebAPIException const& we) { mf::LogPrint("BNBRetriever") << "At time " << time << " '" << devName << "' got exception: " << we.what(); }
        };
      fillFromDev(times_temps[i], "E:TOR860@",&TOR860,&TOR860_time);
```
or (always returning the first value as a `double`):
```suggestion
      auto readFromDev = [this](double time, std::string const& varList, auto&... others) -> double
        {
          double v = 0.0;
          try { bfp->GetNamedData(time, varList, v, &others...); }
          catch (WebAPIException const& we) { mf::LogPrint("BNBRetriever") << "At time " << time << " '" << devName << "' got exception: " << we.what(); }
          return v;
        };
      double TOR860 = readFromDev(times_temps[i], "E:TOR860@",&TOR860_time);
```